Improved walk animation and stand animation

An0n3m0us opened the pull request:
I've improved the walk animation by adding bobbing to it:
Old: https://my.pcloud.com/publink/show?code=XZQw2A7ZgJflKzKLHQzEtvmo1BehpVYxiSHy
New: https://my.pcloud.com/publink/show?code=XZuw2A7ZXzoItkALCJmWnc6lJUxseLjmFExV

Second commit:
Stand animation issue (closed?): https://github.com/minetest/minetest_game/issues/2220
(I made a second commit just in case we don't want this change)
https://my.pcloud.com/publink/show?code=XZOw2A7Z1Qg1iNYd1Dy1JPfRCYLK7mxKHAXk
Old ---> New
The 'distance moved' for the bones is 1/2 of what it used to be.

LoneWolfHT commented:
I tried this out in singleplayer. It's a nice addition :+1:

paramat commented:
The standing animation is better.
See #2220 can you half the speed of the animation too?
I will +1 this.

For walking, i like the vertical bobbing but dislike the sideways tilting. People don't tilt sideways when they walk. I know that view bobbing has a sideways tilting but i think that is a mistake.
However wait for more core dev input before making changes.

An0n3m0us commented:
Sideways tilting was there already.

paramat commented:
> Sideways tilting was there already

Oops sorry, ok i'll test this it might be good.

An0n3m0us commented:
One moment, you are right. It does look better with `player_set_animation(player, "stand", model.animation_speed / 2)`

paramat commented:
The standing animation still has the same animation speed, sorry, by speed i meant anumation speed not speed of body movement.

An0n3m0us commented:
Animation speed:
https://my.pcloud.com/publink/show?code=XZ2GWA7ZQUGNgQiqkOhzlPmJTr265HCApAD7
Normal ---> 1/2
Which one is better?

paramat commented:
The slower, half speed has support from SmallJoker (core dev) in #2220

An0n3m0us commented:
Ok, I'll push a commit then.

paramat commented:
I'll test all this and will probably +1.

An0n3m0us commented:
https://github.com/minetest/minetest_game/pull/2377
@paramat Any idea what I should do about that pull request? The checks seem to have failed. Shall I just leave it or create a new pull request?

SmallJoker commented:
Half animation speed looks better than full speed, although also a value in between (0.75) might be satisfying in case someone thinks it's too slow with 0.5.
@An0n3m0us I pointed out the issues in that PR. If you have similar questions, you might want to ask on #minetest-dev (IRC, freenode).

An0n3m0us commented:
I tried 3/4, but I think it's a little too fast. 1/2 looks better.

paramat commented on mods/player_api/api.lua:
```diff
@@ -132,7 +132,7 @@ minetest.register_globalstep(function(dtime)
 				player_set_animation(player, "mine")
 			else
-				player_set_animation(player, "stand", animation_speed_mod)
+				player_set_animation(player, "stand", model.animation_speed / 2)
```
We can't do this, as other player models would have their stand animation speed broken.
The standing animation speed would have to be halved in the model itself.
So, i suggest removing this `/ 2` as the rest of the PR is fone.

An0n3m0us replied:
Will do. Other player models? Like in mods?

paramat replied:
Yes, it's an API so any model can be registered for the player.

paramat commented:
In testing, i notice that the vertical movement has different amplitudes for the 2 strides, looks like a limp. This needs fixing.
However, the larger vertical 'bob' on one of the strides looks good.

An0n3m0us commented:
Looks fine to me:
https://my.pcloud.com/publink/show?code=XZbHJN7ZKxPHJe41ab4jJDN0UbLo3HVv6XBV

Yeah, the second stride is the first inversed because using the same stride on the first looks odd on the second.

An0n3m0us commented:
Do you mean in the walk animation or the walk_mine animation? If the former, then it looks fine. Both strides have the exact same amplitude (-0.15 to 0 to 0.15):
https://my.pcloud.com/publink/show?code=XZbHJN7ZKxPHJe41ab4jJDN0UbLo3HVv6XBV

An0n3m0us commented:
I fixed the standing animation using the model instead of the code. Same animation speed and length, but I changed the movement to make it less prominent:
https://my.pcloud.com/publink/show?code=XZhuJN7ZcjgAXqIhLf0UMa35XOwqfSk1wJfk
(New ----> Old)

An0n3m0us commented:
Fixed sitting animation too. Looked like he was hyperventilating xD:
https://my.pcloud.com/publink/show?code=XZHgJN7ZmcBiRX6AsOmzLvzB6ItRizqnCB2X
(New ---> Old)

paramat commented:
> Do you mean in the walk animation

Yes. Hmm this is odd. I use F7 to view the front of the player. Maybe i'm imagining it, will test again.

paramat commented:
> Do you mean in the walk animation

Yes.
The difference in upwards bob is clearly visible in that blender video you linked.
Watch the belt clip, the upwards bob is larger for left leg forwards.

Not sure about the new stand animation, the reduction in body movement is good but now there's a lot of head nodding with a slightly awkward and non-smooth movement.

The sit animation is much better, but there's even more head nodding.

Personally i would prefer the stand / sit animations to have the same motions, just with less amplitude.

paramat commented:
Checked the walk / mine animation, this has very uneven bobbing on the strides, looks bad.

An0n3m0us commented:
> The difference in upwards bob is clearly visible in that blender video you linked.
> Watch the belt clip, the upwards bob is larger for left leg forwards.

That's an illusion from the perspective ; ) I'll check that anyway later.

> Not sure about the new stand animation, the reduction in body movement is good but now there's a > lot of head nodding with a slightly awkward and non-smooth movement.
xD that was already there. Maybe I need to slow that down too, because I didn't do anything to it yet.

> The sit animation is much better, but there's even more head nodding.
Same as before. I'll look into that too.

> Personally i would prefer the stand / sit animations to have the same motions, just with less amplitude.
Got it!

> Checked the walk / mine animation, this has very uneven bobbing on the strides, looks bad.
Yeah, I think I still need to work on that one.

paramat commented:
> That's an illusion from the perspective

It isn't, because i checked by viewing the player directly from the front.
Doing that, press 'Z' for zoom and watch the player's neckline relative to the hotbar.

An0n3m0us commented:
Sorry, I got confused. _Now_ I see what you mean.

Animation:
```
0      5     10     15    20
0    -0.3    0.3   -0.3   0
```

The jump is 0.6 from frames 5 to 10 but only ±0.3 for the rest.

I've now changed it to:
```
0      5     10     15    20
0    -0.3    0     -0.3   0
```

An0n3m0us commented:
I fixed the players' head movement in the standing and sitting animation:
https://my.pcloud.com/publink/show?code=XZXAFN7ZQ2jMVbpsWShJtNzvtD3l95ce4n67
https://my.pcloud.com/publink/show?code=XZ0AFN7Z6ozhmfa1QHXNdrSn9Gj88zcLriGy

Pushing a commit soon.

paramat commented:
Walk, walk/mine, stand, sit are all good now :+1:

An0n3m0us commented:
Done!

paramat commented:
As for #2220 , slower (half speed) standing and sitting animations would be ideal. Can that be done in the model itself?
Doing so may break mods that use certain frames of the model for emotes etc. but they could update.

paramat commented on mods/player_api/README.txt:
```diff
@@ -30,3 +30,6 @@ celeron55, Perttu Ahola <celeron55@gmail.com> (CC BY-SA 3.0):
 https://www.freesound.org/people/sonictechtonic/sounds/241872/
   player_damage.ogg
+
+An0n3m0us (CC BY 3.0):
```
This is best moved to be under:
```
stujones11 (CC BY-SA 3.0):
character.b3d
character.blend -- Both derived from a model by MirceaKitsune (CC BY-SA 3.0)
```
And your license has to be the same license because you have altered something licensed as 'share alike'.

An0n3m0us replied:
Will do.

>And your license has to be the same license because you have altered something licensed as 'share alike'.

Oh, I meant to do that but I forgot.

An0n3m0us replied:
Like this?:
```
stujones11 (CC BY-SA 3.0):
  character.b3d
  character.blend -- Both derived from a model by MirceaKitsune (CC BY-SA 3.0)

An0n3m0us (CC BY-SA 3.0):
  character.b3d
  character.blend -- Player animation improvement
```

paramat replied:
That's fine for me.

An0n3m0us replied:
OK, done.

paramat commented on mods/player_api/README.txt:
```diff
@@ -30,3 +30,6 @@ celeron55, Perttu Ahola <celeron55@gmail.com> (CC BY-SA 3.0):
   player_damage.ogg
+
+An0n3m0us (CC BY 3.0):
+  character.blend -- Player animation improvement
```
`character.b3d` too as you have altered that?

An0n3m0us replied:
Oh yeah, of course xD . I'll do that too.

An0n3m0us replied:
Done.

paramat commented:
:+1: 
Perhaps we can deal with slower movement later as it will be more difficult and may have some issues.

An0n3m0us commented:
>As for #2220 , slower (half speed) standing and sitting animations would be ideal. Can that be done in >the model itself?
>Doing so may break mods that use certain frames of the model for emotes etc. but they could update.

The movement can be modified to be faster or slower quite easily without adding anymore or less frames and without moving them around either.

paramat commented:
> Movement can be 'slowed' without changing the frames by decreasing the distance moved by the object. That's what I did previously for the sitting and stand animation.

Sorry, as before, by 'slower' i mean a lower frequency for the 'breathing' movement, just as 'animation speed / 2' does. I don't mean the same frequency but lower velocity of movement, which your changes already do.

An0n3m0us commented:
Currently we have two occurences (frequencies) for the stand ('breathing') animation. You said:
> lower frequency for the 'breathing' movement

So you want only one occurence of the 'breathing' animation?

paramat commented:
> I think I get what you mean. You mean to say that the timespan needs to be expanded so that the movement of the body per frame is less (technically 'slower') as it has a longer timespan to progress through, right?
Like this

Yes like that. This is what #2220 is requesting.
However, let's wait for more core dev input before attempting this, as it would cause some mod breakage.
///////////////////////////////
> Currently we have two occurences (frequencies)

Frequency isn't number of events.

> And animation speed / 2 wouldn't reduce the frequency either. It would just slow down the playback.

Animation speed / 2 does reduce the frequency.
By frequency i mean the scientific definition "number of events per second".

> But animation speed / 2 slows the playback speed and the frequency is still the same isn't it?

No.

> Now if we were to play the animation at a speed of speed/2, the playback will still go through 0-60 which means it will play two jumps.

Yes, but in a time period twice as long, so frequency has halved.
frequency = number of events / time period in seconds

An0n3m0us commented:
>  However, let's wait for more core dev input before attempting this, as it would cause some mod breakage.

:+1: I agree.

> By frequency i mean the scientific definition "number of events per second".
> but in a time period twice as long, so frequency has halved
> frequency = number of events / time period in seconds

Oh, I see. I was thinking of the other type (absolute frequency), as in the amount of repeats (occurences) in any given timespan. `Relative frequency` is where you divide the frequency by the total number of data values (in this case, time).

TumeniNodes commented:
Hard to get an idea of the examples when the links are all  deleted : /

An0n3m0us commented:
Yeah, I removed the shares (but kept the files) because I didn't want unnecessary traffic wastage.

TumeniNodes commented:
The new animation looks much better.
I swear I heard the BeeGees, Saturday Night Fever playing when I watched it.
But there is talk of switching to a new model https://github.com/minetest/minetest_game/issues/1235#issuecomment-504772498

An0n3m0us commented:
Must've been your imagination xD Anyway, I had audio set off.

Why, there isn't a need for a new model! This one works/looks fine.
Also, it's possible to use the current model with a 64x64 skin. It's a matter of UV coordinates, not the model itself. I won't comment much on that topic because I haven't read the full thing (it's too long!).

paramat commented:
Don't worry about the 1.8 skin thing, if it goes ahead it won't be soon, this PR should still be merged.

An0n3m0us commented:
I hope this is merged soon otherwise it could cause irreparable merge conflict if someone was to create a pull request in which they've changed the player model that I've already changed.
As for the sneak animation from #2294, that doesn't really matter as I've added new frames instead of modifying original frames.

paramat commented:
Before i approve i'd like @stujones11 or another modelling expert to check this has been done correctly. I don't do modelling so i can't thoroughly check this.

An0n3m0us commented:
Just to be clear, I didn't change the model at all in any way so it's not "modelling" but "animation". I modified the walking animation of the player.

paramat commented:
Ok, i mean 'animation' then. But still, i cannot review this, we need someone who can check the animation in Blender.

NathanSalapat commented:
I don't know if I count as an expert, but I checked the model out, everything looks good, both in Blender and in the game.

paramat commented:
NathanSalapat thanks, that is good enough for me.

An0n3m0us commented:
@NathanSalapat Thanks for checking it : )

paramat commented:
I will re-review this and will probably +1, i was close to doing so before.

SmallJoker commented:
playeranim (https://github.com/minetest-mods/playeranim) now looks like this:

![grafik](https://user-images.githubusercontent.com/1497498/71284237-94e4a100-2362-11ea-9fc5-234f733dbdc8.png)

An0n3m0us commented:
Odd. I'll check it out.

An0n3m0us commented:
OK, I found the cause; in the master branch, the players animation starts at 0, but for some reason when I animated the model the animation started at 1.

An0n3m0us commented:
Sorry about those force-pushes; I'm fighting with the license file and Git; they're winning so far . . .

SmallJoker commented:
:+1: tested, works.

paramat commented:
Tested again and seems ok :+1:

paramat commented:
Thanks for your patience and perseverance.

paramat commented:
@An0n3m0us 
I cannot remember if this PR only reduces the amplitude of the standing animation, or actually reduces the frequency, or both?

An0n3m0us commented:
The frequency is still the same; only the amplitude is decreased for both standing and sitting animation.

paramat commented:
Thanks.
